From 5a4a6b2be672796f837d70f2bbe51c1353f576f7 Mon Sep 17 00:00:00 2001
From: Reilly Brogan <reilly@reillybrogan.com>
Date: Sun, 7 Sep 2025 15:15:45 -0500
Subject: [PATCH] Backport ffmpeg 8.0 build fixes to 25.08 branch

Building with ffmpeg 8.0 was fixed in d39fae995420d5b78d40a51a5a748c1ecd74a0ee by @amazingakai however that commit also raised the minimum ffmpeg required to build ffmpegthumbs. As we can't increase version requirements in release branches that commit cannot be cherry-picked as-is.

This commit instead takes the relevant changes from that commit and wraps them in version check ifdefs. That allows us to fix the build with ffmpeg 8.0 but without breaking compatibility with older ffmpeg version. This code is ugly but it shouldn't matter since it will never hit any branch besides the 25.08 one.

LIBAVCODEC_VERSION_MAJOR 60 corresponds to ffmpeg 6.x for the record.
---
 ffmpegthumbnailer/moviedecoder.cpp | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/ffmpegthumbnailer/moviedecoder.cpp b/ffmpegthumbnailer/moviedecoder.cpp
index 39a7c00..e6e5dd4 100644
--- a/ffmpegthumbnailer/moviedecoder.cpp
+++ b/ffmpegthumbnailer/moviedecoder.cpp
@@ -12,6 +12,7 @@
 
 extern "C" {
 #include <libswscale/swscale.h>
+#include <libavcodec/version.h>
 #include <libavutil/display.h>
 #include <libavutil/imgutils.h>
 }
@@ -89,7 +90,11 @@ void MovieDecoder::destroy()
 {
     deleteFilterGraph();
     if (m_pVideoCodecContext) {
+#if LIBAVCODEC_VERSION_MAJOR >= 60
+        avcodec_free_context(&m_pVideoCodecContext);
+#else
         avcodec_close(m_pVideoCodecContext);
+#endif
         m_pVideoCodecContext = nullptr;
     }
     m_pVideoStream = nullptr;
@@ -215,7 +220,11 @@ void MovieDecoder::seek(int timeInSeconds)
         }
 
         ++keyFrameAttempts;
+#if LIBAVCODEC_VERSION_MAJOR >= 60
+    } while ((!gotFrame || !(m_pFrame->flags & AV_FRAME_FLAG_KEY)) && keyFrameAttempts < 200);
+#else
     } while ((!gotFrame || !m_pFrame->key_frame) && keyFrameAttempts < 200);
+#endif
 
     if (gotFrame == 0) {
         qCDebug(ffmpegthumbs_LOG) << "Seeking in video failed";
@@ -263,15 +272,29 @@ QImageIOHandler::Transformations MovieDecoder::transformations()
         return ret;
     }
 
+#if LIBAVCODEC_VERSION_MAJOR >= 60
+    for (int i=0; i<m_pVideoStream->codecpar->nb_coded_side_data; i++) {
+        if (m_pVideoStream->codecpar->coded_side_data[i].type != AV_PKT_DATA_DISPLAYMATRIX) {
+#else
     for (int i=0; i<m_pVideoStream->nb_side_data; i++) {
         if (m_pVideoStream->side_data[i].type != AV_PKT_DATA_DISPLAYMATRIX) {
+#endif
             continue;
         }
+#if LIBAVCODEC_VERSION_MAJOR >= 60
+        if (m_pVideoStream->codecpar->coded_side_data[i].size != sizeof(int32_t) * 9) {
+            qCWarning(ffmpegthumbs_LOG) << "Invalid display matrix size" << m_pVideoStream->codecpar->coded_side_data[i].size << "expected" << sizeof(int32_t) * 9;
+#else
         if (m_pVideoStream->side_data[i].size != sizeof(int32_t) * 9) {
             qCWarning(ffmpegthumbs_LOG) << "Invalid display matrix size" << m_pVideoStream->side_data[i].size << "expected" << sizeof(int32_t) * 9;
+#endif
             continue;
         }
+#if LIBAVCODEC_VERSION_MAJOR >= 60
+        int32_t *matrix = reinterpret_cast<int32_t*>(m_pVideoStream->codecpar->coded_side_data[i].data);
+#else
         int32_t *matrix = reinterpret_cast<int32_t*>(m_pVideoStream->side_data[i].data);
+#endif
         double rotation = av_display_rotation_get(matrix);
         if (qFuzzyCompare(rotation, 0.)) {
             ret |= QImageIOHandler::TransformationNone;
@@ -404,7 +427,11 @@ bool MovieDecoder::processFilterGraph(AVFrame *dst, const AVFrame *src,
 
 void MovieDecoder::getScaledVideoFrame(int scaledSize, bool maintainAspectRatio, VideoFrame& videoFrame)
 {
+#if LIBAVCODEC_VERSION_MAJOR >= 60
+    if (m_pFrame->flags & AV_FRAME_FLAG_INTERLACED) {
+#else
     if (m_pFrame->interlaced_frame) {
+#endif
         processFilterGraph((AVFrame*) m_pFrame, (AVFrame*) m_pFrame, m_pVideoCodecContext->pix_fmt,
                               m_pVideoCodecContext->width, m_pVideoCodecContext->height);
     }
-- 
2.51.0

